File: apps/sandbox-runtime/internal/egress/forwarder.go
package egress

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/startup"
)

const (
	HeaderEgressRouteID               = "X-Mistle-Egress-Route-Id"
	HeaderEgressBindingID             = "X-Mistle-Egress-Binding-Id"
	HeaderEgressUpstreamBaseURL       = "X-Mistle-Egress-Upstream-Base-Url"
	HeaderEgressAuthInjectionType     = "X-Mistle-Egress-Auth-Injection-Type"
	HeaderEgressAuthInjectionTarget   = "X-Mistle-Egress-Auth-Injection-Target"
	HeaderEgressAuthInjectionUsername = "X-Mistle-Egress-Auth-Injection-Username"
	HeaderEgressConnectionID          = "X-Mistle-Egress-Connection-Id"
	HeaderEgressCredentialSecretType  = "X-Mistle-Egress-Credential-Secret-Type"
	HeaderEgressCredentialPurpose     = "X-Mistle-Egress-Credential-Purpose"
	HeaderEgressCredentialResolverKey = "X-Mistle-Egress-Credential-Resolver-Key"
	HeaderSandboxProfileID            = "X-Mistle-Sandbox-Profile-Id"
	HeaderSandboxProfileVersion       = "X-Mistle-Sandbox-Profile-Version"
)

type Forwarder struct {
	httpClient            *http.Client
	tokenizerProxyBaseURL *url.URL
	runtimePlan           startup.RuntimePlan
}

type NewForwarderInput struct {
	HTTPClient                  *http.Client
	TokenizerProxyEgressBaseURL string
	RuntimePlan                 startup.RuntimePlan
}

func NewForwarder(input NewForwarderInput) (Forwarder, error) {
	if input.HTTPClient == nil {
		return Forwarder{}, fmt.Errorf("http client is required")
	}

	tokenizerProxyBaseURL, err := url.Parse(strings.TrimSpace(input.TokenizerProxyEgressBaseURL))
	if err != nil {
		return Forwarder{}, fmt.Errorf("failed to parse tokenizer proxy egress base url: %w", err)
	}
	if tokenizerProxyBaseURL.Scheme != "http" && tokenizerProxyBaseURL.Scheme != "https" {
		return Forwarder{}, fmt.Errorf("tokenizer proxy egress base url must use http or https scheme")
	}
	if tokenizerProxyBaseURL.Host == "" {
		return Forwarder{}, fmt.Errorf("tokenizer proxy egress base url host is required")
	}

	return Forwarder{
		httpClient:            input.HTTPClient,
		tokenizerProxyBaseURL: tokenizerProxyBaseURL,
		runtimePlan:           input.RuntimePlan,
	}, nil
}

func joinPath(basePath string, suffixPath string) string {
	normalizedBasePath := strings.TrimSuffix(basePath, "/")
	normalizedSuffixPath := strings.TrimPrefix(suffixPath, "/")

	if normalizedBasePath == "" {
		normalizedBasePath = "/"
	}

	if normalizedSuffixPath == "" {
		return normalizedBasePath
	}

	if normalizedBasePath == "/" {
		return "/" + normalizedSuffixPath
	}

	return normalizedBasePath + "/" + normalizedSuffixPath
}

func createTokenizerProxyURL(input struct {
	baseURL    *url.URL
	targetPath string
	rawQuery   string
}) string {
	forwardURL := *input.baseURL
	forwardURL.Path = joinPath(forwardURL.Path, input.targetPath)
	forwardURL.RawQuery = input.rawQuery
	forwardURL.Fragment = ""
	return forwardURL.String()
}

func copyHeadersWithoutHopByHop(target http.Header, source http.Header) {
	for headerName, values := range source {
		normalizedHeaderName := strings.ToLower(headerName)
		switch normalizedHeaderName {
		case "connection", "proxy-connection", "keep-alive", "proxy-authenticate", "proxy-authorization", "te", "trailer", "transfer-encoding", "upgrade", "host":
			continue
		}

		for _, value := range values {
			target.Add(headerName, value)
		}
	}
}

func (forwarder Forwarder) buildForwardRequest(input struct {
	incomingRequest *http.Request
	route           startup.EgressCredentialRoute
	targetPath      string
}) (*http.Request, error) {
	forwardRequestURL := createTokenizerProxyURL(struct {
		baseURL    *url.URL
		targetPath string
		rawQuery   string
	}{
		baseURL:    forwarder.tokenizerProxyBaseURL,
		targetPath: input.targetPath,
		rawQuery:   input.incomingRequest.URL.RawQuery,
	})

	forwardRequest, err := http.NewRequestWithContext(
		input.incomingRequest.Context(),
		input.incomingRequest.Method,
		forwardRequestURL,
		input.incomingRequest.Body,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create tokenizer proxy request: %w", err)
	}

	copyHeadersWithoutHopByHop(forwardRequest.Header, input.incomingRequest.Header)
	forwardRequest.Header.Set(HeaderEgressRouteID, input.route.RouteID)
	forwardRequest.Header.Set(HeaderEgressBindingID, input.route.BindingID)
	forwardRequest.Header.Set(HeaderEgressUpstreamBaseURL, input.route.Upstream.BaseURL)
	forwardRequest.Header.Set(HeaderEgressAuthInjectionType, input.route.AuthInjection.Type)
	forwardRequest.Header.Set(HeaderEgressAuthInjectionTarget, input.route.AuthInjection.Target)
	if strings.TrimSpace(input.route.AuthInjection.Username) != "" {
		forwardRequest.Header.Set(
			HeaderEgressAuthInjectionUsername,
			input.route.AuthInjection.Username,
		)
	}
	forwardRequest.Header.Set(HeaderEgressConnectionID, input.route.CredentialResolver.ConnectionID)
	forwardRequest.Header.Set(HeaderEgressCredentialSecretType, input.route.CredentialResolver.SecretType)
	if strings.TrimSpace(input.route.CredentialResolver.Purpose) != "" {
		forwardRequest.Header.Set(
			HeaderEgressCredentialPurpose,
			input.route.CredentialResolver.Purpose,
		)
	}
	if strings.TrimSpace(input.route.CredentialResolver.ResolverKey) != "" {
		forwardRequest.Header.Set(
			HeaderEgressCredentialResolverKey,
			input.route.CredentialResolver.ResolverKey,
		)
	}
	forwardRequest.Header.Set(HeaderSandboxProfileID, forwarder.runtimePlan.SandboxProfileID)
	forwardRequest.Header.Set(
		HeaderSandboxProfileVersion,
		fmt.Sprintf("%d", forwarder.runtimePlan.Version),
	)

	return forwardRequest, nil
}

func (forwarder Forwarder) Forward(input struct {
	incomingRequest *http.Request
	route           startup.EgressCredentialRoute
	targetPath      string
}) (*http.Response, error) {
	forwardRequest, err := forwarder.buildForwardRequest(struct {
		incomingRequest *http.Request
		route           startup.EgressCredentialRoute
		targetPath      string
	}{
		incomingRequest: input.incomingRequest,
		route:           input.route,
		targetPath:      input.targetPath,
	})
	if err != nil {
		return nil, err
	}

	response, err := forwarder.httpClient.Do(forwardRequest)
	if err != nil {
		return nil, fmt.Errorf("failed to forward request to tokenizer proxy: %w", err)
	}

	return response, nil
}

File: apps/sandbox-runtime-node/src/runtime/proxy/forwarder.ts
import type { IncomingHttpHeaders, OutgoingHttpHeaders } from "node:http";

import type { EgressCredentialRoute } from "@mistle/integrations-core";

import {
  copyHeadersWithoutHopByHop,
  headerBagFromIncomingHeaders,
  headerBagToOutgoingHeaders,
  type HeaderBag,
} from "./headers.js";

export const HEADER_EGRESS_RULE_ID = "x-mistle-egress-rule-id";
export const HEADER_EGRESS_BINDING_ID = "x-mistle-egress-binding-id";
export const HEADER_EGRESS_UPSTREAM_BASE_URL = "x-mistle-egress-upstream-base-url";
export const HEADER_EGRESS_AUTH_INJECTION_TYPE = "x-mistle-egress-auth-injection-type";
export const HEADER_EGRESS_AUTH_INJECTION_TARGET = "x-mistle-egress-auth-injection-target";
export const HEADER_EGRESS_AUTH_INJECTION_USERNAME = "x-mistle-egress-auth-injection-username";
export const HEADER_EGRESS_CONNECTION_ID = "x-mistle-egress-connection-id";
export const HEADER_EGRESS_CREDENTIAL_SECRET_TYPE = "x-mistle-egress-credential-secret-type";
export const HEADER_EGRESS_CREDENTIAL_PURPOSE = "x-mistle-egress-credential-purpose";
export const HEADER_EGRESS_CREDENTIAL_RESOLVER_KEY = "x-mistle-egress-credential-resolver-key";
export const HEADER_SANDBOX_PROFILE_ID = "x-mistle-sandbox-profile-id";
export const HEADER_SANDBOX_PROFILE_VERSION = "x-mistle-sandbox-profile-version";

function joinPath(basePath: string, suffixPath: string): string {
  const normalizedBasePath = basePath.endsWith("/") ? basePath.slice(0, -1) : basePath;
  const normalizedSuffixPath = suffixPath.startsWith("/") ? suffixPath.slice(1) : suffixPath;

  if (normalizedBasePath.length === 0 || normalizedBasePath === "/") {
    return normalizedSuffixPath.length === 0 ? "/" : `/${normalizedSuffixPath}`;
  }

  return normalizedSuffixPath.length === 0
    ? normalizedBasePath
    : `${normalizedBasePath}/${normalizedSuffixPath}`;
}

export type BuildTokenizerProxyRequestInput = {
  tokenizerProxyEgressBaseUrl: string;
  runtimePlan: {
    sandboxProfileId: string;
    version: number;
  };
  route: EgressCredentialRoute;
  targetPath: string;
  rawQuery: string;
  method: string;
  headers: IncomingHttpHeaders;
  body: ReadableStream<Uint8Array> | undefined;
};

export type TokenizerProxyRequest = {
  url: URL;
  method: string;
  headers: OutgoingHttpHeaders;
  body: ReadableStream<Uint8Array> | undefined;
};

function setHeader(
  headerBag: HeaderBag,
  headerName: string,
  headerValue: string | undefined,
): void {
  if (headerValue === undefined || headerValue.trim().length === 0) {
    return;
  }

  headerBag.set(headerName, [headerValue]);
}

export function buildTokenizerProxyRequest(
  input: BuildTokenizerProxyRequestInput,
): TokenizerProxyRequest {
  const tokenizerProxyUrl = new URL(input.tokenizerProxyEgressBaseUrl);
  tokenizerProxyUrl.pathname = joinPath(tokenizerProxyUrl.pathname, input.targetPath);
  tokenizerProxyUrl.search = input.rawQuery.length === 0 ? "" : `?${input.rawQuery}`;
  tokenizerProxyUrl.hash = "";

  const headers = copyHeadersWithoutHopByHop(headerBagFromIncomingHeaders(input.headers), true);
  setHeader(headers, HEADER_EGRESS_RULE_ID, input.route.egressRuleId);
  setHeader(headers, HEADER_EGRESS_BINDING_ID, input.route.bindingId);
  setHeader(headers, HEADER_EGRESS_UPSTREAM_BASE_URL, input.route.upstream.baseUrl);
  setHeader(headers, HEADER_EGRESS_AUTH_INJECTION_TYPE, input.route.authInjection.type);
  setHeader(headers, HEADER_EGRESS_AUTH_INJECTION_TARGET, input.route.authInjection.target);
  setHeader(headers, HEADER_EGRESS_AUTH_INJECTION_USERNAME, input.route.authInjection.username);
  setHeader(headers, HEADER_EGRESS_CONNECTION_ID, input.route.credentialResolver.connectionId);
  setHeader(
    headers,
    HEADER_EGRESS_CREDENTIAL_SECRET_TYPE,
    input.route.credentialResolver.secretType,
  );
  setHeader(headers, HEADER_EGRESS_CREDENTIAL_PURPOSE, input.route.credentialResolver.purpose);
  setHeader(
    headers,
    HEADER_EGRESS_CREDENTIAL_RESOLVER_KEY,
    input.route.credentialResolver.resolverKey,
  );
  setHeader(headers, HEADER_SANDBOX_PROFILE_ID, input.runtimePlan.sandboxProfileId);
  setHeader(headers, HEADER_SANDBOX_PROFILE_VERSION, String(input.runtimePlan.version));

  return {
    url: tokenizerProxyUrl,
    method: input.method,
    headers: headerBagToOutgoingHeaders(headers),
    body: input.body,
  };
}
